PiwikTracker/Queue.swift

```diff
@@ -0,0 +1,24 @@
+import Foundation
+
+/// A queue in which Objects that conform to queueable can be stored.
+protocol Queue {
+    /// The number of queud items.
+    var itemCount: Int { get }
+    
+    /// Queues an Item.
+    ///
+    /// - Parameters:
+    ///   - item: The Item to queue.
+    ///   - completion: A closure to be called once queueing is completed.
+    func queue(item: NSCoding, completion: ()->())
+    
+    /// Dequeues and returns a certain amount of items.
+    ///
+    /// - Parameters:
+    ///   - limit: The maximum amount of items to dequeue. May return less if less are queued.
+    ///   - completion: A closure to be called once the elements are dequeued. The closure will be called with the dequeued items.
+    func dequeue(withLimit limit: Int, completion: (_ items: [Any])->())
```

Comment:
You could even skip the type constraint completely and only do
````
associatedtype T
````
And let implementing types add any type constraints. E.g. in-memory implementations does not need any constraints but a persistent queue could require confirmation to `NSCoding`.

But the current limitation  does not matter much since the implementation will not be reused outside the framework.